src/nodes/NodeConnectionManager.ts

```diff
@@ -579,69 +582,102 @@ class NodeConnectionManager {
+                lastUpdated: 0, // FIXME?
+              },
+            ]);
+          }
+        });
+        return nodes;
+      },
+      timer,
+    );
   }
 
   /**
    * Perform an initial database synchronisation: get the k closest nodes
    * from each seed node and add them to this database
-   * For now, we also attempt to establish a connection to each of them.
-   * If these nodes are offline, this will impose a performance penalty,
-   * so we should investigate performing this in the background if possible.
-   * Alternatively, we can also just add the nodes to our database without
-   * establishing connection.
-   * This has been removed from start() as there's a chicken-egg scenario
-   * where we require the NodeGraph instance to be created in order to get
-   * connections.
+   * Establish a proxy connection to each node before adding it
+   * By default this operation is blocking, set `block` to false to make it
+   * non-blocking
    */
   @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
-  public async syncNodeGraph() {
+  public async syncNodeGraph(block: boolean = true, timer?: Timer) {
     for (const seedNodeId of this.getSeedNodes()) {
       // Check if the connection is viable
       try {
-        await this.getConnection(seedNodeId);
+        await this.getConnection(seedNodeId, timer);
       } catch (e) {
         if (e instanceof nodesErrors.ErrorNodeConnectionTimeout) continue;
         throw e;
       }
-
       const nodes = await this.getRemoteNodeClosestNodes(
         seedNodeId,
         this.keyManager.getNodeId(),
+        timer,
       );
-      for (const n of nodes) {
-        await this.nodeGraph.setNode(n.id, n.address);
+      for (const [nodeId, nodeData] of nodes) {
+        if (!block) {
+          this.queue.push(() =>
+            this.nodeManager!.setNode(nodeId, nodeData.address),
+          );
+        } else {
+          try {
+            await this.nodeManager?.setNode(nodeId, nodeData.address);
+          } catch (e) {
+            if (!(e instanceof nodesErrors.ErrorNodeGraphSameNodeId)) throw e;
+          }
+        }
+      }
+      // Refreshing every bucket above the closest node
+      if (!block) {
+        this.queue.push(async () => {
+          const [closestNode] = (
+            await this.nodeGraph.getClosestNodes(this.keyManager.getNodeId(), 1)
+          ).pop()!;
+          const [bucketIndex] = this.nodeGraph.bucketIndex(closestNode);
+          for (let i = bucketIndex; i < this.nodeGraph.nodeIdBits; i++) {
+            this.nodeManager?.refreshBucketQueueAdd(i);
+          }
+        });
+      } else {
+        const [closestNode] = (
+          await this.nodeGraph.getClosestNodes(this.keyManager.getNodeId(), 1)
+        ).pop()!;
+        const [bucketIndex] = this.nodeGraph.bucketIndex(closestNode);
+        for (let i = bucketIndex; i < this.nodeGraph.nodeIdBits; i++) {
+          this.nodeManager?.refreshBucketQueueAdd(i);
+        }
       }
```

Comment:
We don't really need to duplicate the code here. We can just create one lambda and use that,
```ts
const thing = async () => {
  // do the thing
}
if (!block) {
  this.queue.push(thing);
} else {
  await thing();
}
```